lib/Backends/OpenCL/kernels_fwd_quantized_conv.cl

```diff
@@ -0,0 +1,329 @@
+// v_imsi_0, v_imsi_1 - Spacial dimensions of input.
+// v_imso_0, v_imso_1 - Spacial dimensions of output.
+// v_pad_A - Padding required for tiles of A.
+// v_pad_B - Padding required for tiles of A.
+// workgroup_size_0, workgroup_size_1 - workgroup sizes along each dimension.
+// VWN - vector width in dimension N.
+// VWM - vector width in dimension M.
+
+static const char* FWD_CONV_QUANTIZED_CODE = R"(
+#define Dtype char
+#define Dtype1 char
+#define Dtype2 char2
+#define Dtype4 char4
+#define Dtype8 char8
+#define Dtype16 char16
+
+#define A_OFFSET(x) ((x)- Aoffset)
+#define B_OFFSET(x) ((x)- Boffset)
+
+#define VEC_1_0(X) X
+#define VEC_2_0(X) X.x
+#define VEC_2_1(X) X.y
+#define VEC_4_0(X) X.x
+#define VEC_4_1(X) X.y
+#define VEC_4_2(X) X.z
+#define VEC_4_3(X) X.w
+#define VEC_8_0(X) X.s0
+#define VEC_8_1(X) X.s1
+#define VEC_8_2(X) X.s2
+#define VEC_8_3(X) X.s3
+#define VEC_8_4(X) X.s4
+#define VEC_8_5(X) X.s5
+#define VEC_8_6(X) X.s6
+#define VEC_8_7(X) X.s7
+#define VEC_16_0(X) X.s0
+#define VEC_16_1(X) X.s1
+#define VEC_16_2(X) X.s2
+#define VEC_16_3(X) X.s3
+#define VEC_16_4(X) X.s4
+#define VEC_16_5(X) X.s5
+#define VEC_16_6(X) X.s6
+#define VEC_16_7(X) X.s7
+#define VEC_16_8(X) X.s8
+#define VEC_16_9(X) X.s9
+#define VEC_16_10(X) X.sA
+#define VEC_16_11(X) X.sB
+#define VEC_16_12(X) X.sC
+#define VEC_16_13(X) X.sD
+#define VEC_16_14(X) X.sE
+#define VEC_16_15(X) X.sF
+
+#define int_tp int
+#define uint_tp unsigned int
+#define int_tpc int
+#define uint_tpc unsigned int
+
+// Input image size in pixels.
+#define v_imsi (v_imsi_0*v_imsi_1)
+// Output image size in pixels.
+#define v_imso (v_imso_0*v_imso_1)
+// Input image batch offset.
+#define v_B_off  (v_fin*v_imsi)
+// Output image batch offset.
+#define v_C_off (v_fout*v_imso)
+// Definitions used by the GEMM kernel.
+#define MG v_fout
+#define MM (v_fout/v_g)
+#define NN v_imso
+#define KG (v_fin*v_k_0*v_k_1)
+#define KK ((v_fin/v_g)*v_k_0*v_k_1)
+// The tile-size in dimension M.
+#define TSM (WPTM*workgroup_size_1)
+// The tile-size in dimension N.
+#define TSN (WPTN*workgroup_size_0)
+// The reduced tile-size in dimension M.
+#define RTSM workgroup_size_1
+// The reduced tile-size in dimension N.
+#define RTSN workgroup_size_0
+// Loads per thread for A.
+#ifdef LPTA
+#undef LPTA
+#endif
+#define LPTA ((TSK*TSM)/(RTSM*RTSN))
+// Loads per thread for B.
+#ifdef LPTB
+#undef LPTB
+#endif
+#define LPTB ((TSK*TSN)/(RTSM*RTSN))
+#ifdef v_num_tiles
+#undef v_num_tiles
+#endif
+#define v_num_tiles (((KK - 1)/(TSK*2) + 1)*2)
+
+/// Clips int into char.
+char clip(int val) { return (char)min(max(val, -128), 127); }
+
+__kernel
+__attribute__((reqd_work_group_size(workgroup_size_0, workgroup_size_1, 1)))
+__attribute__((vec_type_hint(Dtype4)))
+void
+conv_forward_mem_i8(__global void *mem, unsigned im_in_offset, unsigned wg_offset,
+                 unsigned bias_offset, unsigned im_out_offset,
+                 int Aoffset, float Ascale, int Boffset, float Bscale,
```

Comment:
The naming of offsets and scales does not match the naming of other parameters, which start lowercase and use underscores.

Reply:
changed